lib/hpke/src/signature.cpp

```diff
@@ -95,69 +101,246 @@ struct ConcreteSignature : public Signature
     return group.verify(data, sig, rpk);
   }
 
-  template<Signature::ID id>
-  static const ConcreteSignature instance;
-
 private:
   const Group& group;
 };
 
-template<>
-const ConcreteSignature
-  ConcreteSignature::instance<Signature::ID::P256_SHA256> =
-    ConcreteSignature(Group::get<Group::ID::P256>());
+// XXX(RLB): There is a lot of code in RSASignature that is duplicated in
+// EVPGroup.  I have allowed this duplication rather than factoring it out
+// because I would like to be able to cleanly remove RSA later.
+struct RSASignature : public Signature
+{
+  struct PublicKey : public Signature::PublicKey
+  {
+    explicit PublicKey(EVP_PKEY* pkey_in)
+      : pkey(pkey_in, typed_delete<EVP_PKEY>)
+    {}
 
-template<>
-const ConcreteSignature
-  ConcreteSignature::instance<Signature::ID::P384_SHA384> =
-    ConcreteSignature(Group::get<Group::ID::P384>());
+    ~PublicKey() override = default;
 
-template<>
-const ConcreteSignature
-  ConcreteSignature::instance<Signature::ID::P521_SHA512> =
-    ConcreteSignature(Group::get<Group::ID::P521>());
+    typed_unique_ptr<EVP_PKEY> pkey;
+  };
 
-template<>
-const ConcreteSignature ConcreteSignature::instance<Signature::ID::Ed25519> =
-  ConcreteSignature(Group::get<Group::ID::Ed25519>());
+  struct PrivateKey : public Signature::PrivateKey
+  {
+    explicit PrivateKey(EVP_PKEY* pkey_in)
+      : pkey(pkey_in, typed_delete<EVP_PKEY>)
+    {}
 
-template<>
-const ConcreteSignature ConcreteSignature::instance<Signature::ID::Ed448> =
-  ConcreteSignature(Group::get<Group::ID::Ed448>());
+    ~PrivateKey() override = default;
+
+    std::unique_ptr<Signature::PublicKey> public_key() const override
+    {
+      if (1 != EVP_PKEY_up_ref(pkey.get())) {
+        throw openssl_error();
+      }
+      return std::make_unique<PublicKey>(pkey.get());
+    }
+
+    typed_unique_ptr<EVP_PKEY> pkey;
+  };
+
+  explicit RSASignature(Digest::ID digest)
+    : Signature(Signature::ID::RSA_SHA256)
```

Comment:
Let's be hash-agnostic here.
```suggestion
    : Signature(digest_to_sig(digest))
```

... where `digest_to_sig()` should just be a `switch` statement like `digest_to_md` -- but with only one arm, for SHA256.